src/apps/wall/scanner.lua

```diff
@@ -0,0 +1,299 @@
+module(..., package.seeall)
+
+local util = require("apps.wall.util")
+local lib  = require("core.lib")
+local bit  = require("bit")
+local ffi  = require("ffi")
+
+local rd16, rd32, ipv6_addr_cmp = util.rd16, util.rd32, util.ipv6_addr_cmp
+local tobit, lshift, rshift = bit.tobit, bit.lshift, bit.rshift
+local band, bxor, bnot = bit.band, bit.bxor, bit.bnot
+
+-- Constants: Ethernet
+ETH_TYPE_IPv4        = lib.htons(0x0800)
+ETH_TYPE_IPv6        = lib.htons(0x86DD)
+ETH_TYPE_VLAN        = lib.htons(0x8100)
+ETH_TYPE_OFFSET      = 12
+ETH_HEADER_SIZE      = 14
+
+-- Constants: IPv4
+IPv4_VER_IHL_OFFSET  = 0
+IPv4_DSCP_ECN_OFFSET = 1
+IPv4_LEN_OFFSET      = 2
+IPv4_FRAG_ID_OFFSET  = 4
+IPv4_FLAGS_OFFSET    = 6
+IPv4_TTL_OFFSET      = 8
+IPv4_PROTO_OFFSET    = 9
+IPv4_CHECKSUM_OFFSET = 10
+IPv4_SRC_ADDR_OFFSET = 12
+IPv4_DST_ADDR_OFFSET = 16
+
+IPv4_PROTO_TCP       = 6   -- uint8_t
+IPv4_PROTO_UDP       = 17  -- uint8_t
+
+-- Constants: IPv6
+IPv6_MIN_HEADER_SIZE = 40
+IPv6_PLOADLEN_OFFSET = 4
+IPv6_NEXTHDR_OFFSET  = 6
+IPv6_HOPLIMIT_OFFSET = 7
+IPv6_SRC_ADDR_OFFSET = 8
+IPv6_DST_ADDR_OFFSET = 26
+
+IPv6_NEXTHDR_HOPBYHOP= 0
+IPv6_NEXTHDR_TCP     = 6
+IPv6_NEXTHDR_UDP     = 17
+IPv6_NEXTHDR_ROUTING = 43
+IPv6_NEXTHDR_FRAGMENT= 44
+IPv6_NEXTHDR_ESP     = 50
+IPv6_NEXTHDR_AH      = 51
+IPv6_NEXTHDR_ICMPv6  = 58
+IPv6_NEXTHDR_NONE    = 59
+IPv6_NEXTHDR_DSTOPTS = 60
+
+-- Constants: TCP
+TCP_HEADER_SIZE      = 20
+TCP_SRC_PORT_OFFSET  = 0
+TCP_DST_PORT_OFFSET  = 2
+
+-- Constants: UDP
+UDP_HEADER_SIZE      = 8
+UDP_SRC_PORT_OFFSET  = 0
+UDP_DST_PORT_OFFSET  = 2
+
+
+ffi.cdef [[
+   struct swall_flow_key_ipv4 {
+      uint16_t vlan_id;
+      uint8_t  __pad;
+      uint8_t  ip_proto;
+      uint32_t lo_addr;
+      uint32_t hi_addr;
+      uint16_t lo_port;
+      uint16_t hi_port;
+   } __attribute__((packed));
+
+   struct swall_flow_key_ipv6 {
+      uint16_t vlan_id;
+      uint8_t  __pad;
+      uint8_t  ip_proto;
+      uint64_t lo_addr;
+      uint64_t hi_addr;
+      uint16_t lo_port;
+      uint16_t hi_port;
+   } __attribute__((packed));
+]]
+
+local INT32_MIN = -0x80000000
+local function hash32(i32)
+   i32 = tobit(i32)
+   i32 = i32 + bnot(lshift(i32, 15))
+   i32 = bxor(i32, (rshift(i32, 10)))
+   i32 = i32 + lshift(i32, 3)
+   i32 = bxor(i32, rshift(i32, 6))
+   i32 = i32 + bnot(lshift(i32, 11))
+   i32 = bxor(i32, rshift(i32, 16))
+
+   -- Unset the low bit, to distinguish valid hashes from HASH_MAX.
+   i32 = lshift(i32, 1)
+   -- Project result to u32 range.
+   return i32 - INT32_MIN
+end
+
+local uint32_ptr = ffi.typeof("uint32_t*")
```

Comment:
Nit: I think the name of the type should be `uint32_ptr_t` to keep it consistent with other definitions of types such as `uint16_ptr_t` (defined below in `util.lua`).